File: client/smart_lock/lib/models/request.dart
class Request {
  final String id;
  final String userId;
  final String type;
  final String status;
  final String location;
  final DateTime timestamp;
  final String? processedBy;
  final DateTime? processedAt;
  final String? notes;
  final Map<String, dynamic>? metadata;

  Request({
    required this.id,
    required this.userId,
    required this.type,
    required this.status,
    required this.location,
    required this.timestamp,
    this.processedBy,
    this.processedAt,
    this.notes,
    this.metadata,
  });

  factory Request.fromJson(Map<String, dynamic> json) {
    return Request(
      id: json['_id'],
      userId: json['userId'],
      type: json['type'],
      status: json['status'],
      location: json['location'],
      timestamp: DateTime.parse(json['timestamp']),
      processedBy: json['processedBy'],
      processedAt: json['processedAt'] != null
          ? DateTime.parse(json['processedAt'])
          : null,
      notes: json['notes'],
      metadata: json['metadata'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'location': location,
      'notes': notes,
      'metadata': metadata,
    };
  }
}

File: server/src/models/Request.js
const mongoose = require('mongoose');

const requestSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  type: {
    type: String,
    enum: ['rfid', 'door_unlock', 'access_card'],
    required: true
  },
  status: {
    type: String,
    enum: ['pending', 'approved', 'denied'],
    default: 'pending'
  },
  location: {
    type: String,
    required: true
  },
  timestamp: {
    type: Date,
    default: Date.now
  },
  processedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  processedAt: Date,
  notes: String,
  metadata: {
    type: Map,
    of: mongoose.Schema.Types.Mixed
  }
}, {
  timestamps: true
});

// Indexes for better query performance
requestSchema.index({ userId: 1, timestamp: -1 });
requestSchema.index({ status: 1, timestamp: -1 });
requestSchema.index({ processedBy: 1, timestamp: -1 });

module.exports = mongoose.model('Request', requestSchema); 